File: Sorting/Bubble-Sort.py
######Simple Approach########
def BubbleSort(arr):
    n = len(arr)
    for i in range(n-1):   #no need to check for last element as it will be already in correct place just after first iteration
        for j in range(n-1-i):  #ith element will be in correct position so, do not compare them
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]   #swapping
    

#####Optimized Approach######
def BubbleSort1(arr):
    n = len(arr)
    for i in range(n-1):
        swapped = False          
        for j in range(n-1-i):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
                swapped = True

        #If no swapping takes place in inner loop, means array is already sorted, so exit the loops
        if swapped == False:          
            break


#Taking the input
array = list(map(int,input("Enter your array (as space separated integers) : ").split()))

#Passing array to the function
BubbleSort1(array)

print(array)

File: C++/Sorting/Bubble-Sort.cpp
#include <iostream>

void Bubble_Sort(int arr[], int n);
void opt_Bubble_Sort(int arr[], int n);
void swap(int* a, int* b);

int main(){

    int arr[50], n;

    std::cout<< "Enter the size of your array : " ;
    std::cin >> n;

    std::cout<< "Enter the elements of your array : " ;
    for(int i=0; i<n; ++i)
        std::cin >>arr[i];

    opt_Bubble_Sort(arr,n);

    std::cout<< "Sorted array : " << std::endl;
    for(int i=0; i<n; ++i)
        std::cout << arr[i] << ",";
        
    return 0;

}

void Bubble_Sort(int arr[], int n){

    for (int i=0; i<n; ++i)
        for (int j=0; j<n-1-i; ++j)
            if (arr[j] > arr[j+1])
                swap(&arr[j], &arr[j+1]);
}

void opt_Bubble_Sort(int arr[], int n){

    for (int i=0; i<n; ++i){

        bool swapped = false;
        for(int j=0; j<n-i; ++j){
            if(arr[j] > arr[j+1]){
                swap(&arr[j], &arr[j+1]);
                swapped = true;
            }
        }
        if (swapped == false)    
            break;
    }
}

void swap(int* a, int* b){

    int temp;
    temp = *a;
    *a = *b;
    *b = temp;
}
